Built the validation side of overwriting config

Alandlt15 opened the pull request:
## Summary
### Validation Logic Overview:
1. Checks if the value wanting to be changed is actually allowed to be changed (otherwise raises KeyError)
2. If true, checks that the value's type matches the expected type (otherwise raises TypeError)
3. If true, checks that the value is in a range (min >= & <= max) (otherwise raises ValueError)
4. Lastly, checks if its a temporary or permanent change

### Tests:
I created unit tests to make sure validation logic worked as expected:

<img width="619" alt="image" src="https://github.com/user-attachments/assets/05b7b2f0-56d0-4a8a-9be5-ba1201a0b586" />


### Board logs:
How I tested on the board, (correct me if I'm wrong or if there's a better way):
1. Ran `make build` in pysquared repo
2. Ran `make install` in v4 repo 
3. From pysquared repo, I copied contents of pysquared dir within artifacts dir
4. Pasted contents inside pysquared dir within lib on the board

I'm not too sure if the tca error is normal in the logs below, however everything else seems to be running fine:
<img width="868" alt="image" src="https://github.com/user-attachments/assets/8dd0f675-d0f7-4a2e-b550-433c41352a71" />

### Final Note:
When one runs `make test` the tests change the config.json file, so I set the permanent updates to change config.test.json for now.

## How was this tested
- [x] Added new unit tests
- [x] Ran code on hardware (screenshots are helpful)
- [ ] Other (Please describe)

Mikefly123 commented:
Hey @Alandlt15 just wanted to check in on if this was still in the review pipeline? Forgot to ask about it during the weekly this last week.

Alandlt15 commented:
Yes! This is still in the review pipeline. Nate told me on Monday he was going to find sometime to review it.

It would help a lot if you're also wanting to get some eyes on this PR too. Thank you!

nateinaction commented:
Apologies for the delay, I held off on reviewing this this weekend because of how it touches state of health. I'm working through that review first and then will switch to this.

hrfarmer reviewed (commented):
Comments in the review are just a few small things I noticed.

Also just want to point out that right now from what I can tell, if two schemas were to have keys with the same name, the logic for finding the key in a schema would fall apart. Not sure if this is something that *needs* to be addressed currently (since there isn't any shared keys, and there probably won't be), but I think it's still something to keep in mind. There could probably be a comment put somewhere so we have it in mind in the future.

Alandlt15 commented on pysquared/config/config.py:
```diff
@@ -34,9 +34,163 @@ def __init__(self, config_path: str) -> None:
+        }
+
+    # validates values from input
+    def validate(self, key: str, value) -> None:
```
Yes, you're correct. This function should only be used inside of the config class. Thank you for pointing this out!

Alandlt15 commented on pysquared/config/config.py:
```diff
@@ -14,13 +14,13 @@
 
 class Config:
     def __init__(self, config_path: str) -> None:
+        self.temp_config_file = "tests/unit/files/config.test.json"
```
I'm a little unsure on how to handle this one here. I changed it to .test.json, because whenever I or anyone runs `make test` it will actually change the .json file and put everything into one line and its kinda annoying. 

Maybe this is fine, but idk if we can also put a note or comment to change it back to the original json file?

hrfarmer replied:
Well for the one line thing, we probably don't want the config files to be losing their formatting on save at all. If you add an indent to `json.dumps()`, like this
```py
f.write(json.dumps(json_data), indent=2)
```
in all the places that the file is being saved, it will preserve file formatting.

For the file getting changed in `make test`, specifically the save and update config tests, the easiest thing I can think of would be to make a copy of the test JSON file when the test starts, and then delete it after the test finishes.

Alandlt15 replied:
I'll test this out, thank you!

Alandlt15 commented:
@hrfarmer thank you for taking the time to review my PR! I will get to work on it when I get a chance. And to your first comment, I can def put a comment as a reminder to prevent ourselves from keys having the same name.

Alandlt15 commented:
Latest updates:

Cleaned up config a little bit by doing `self.config_file = config_path` in the constructor.
Updated private methods by prefacing their name with a `_`.
Running config tests no longer updates the config file.

hrfarmer commented:
@Alandlt15 I changed _validate to validate because I ended up realizing that it would be useful to have in some other places (specifically to use for #244)

However whenever I try to update the config on the board I'm getting a Read-only filesystem error, just wondering if anyone else is running into this?
<img width="699" alt="image" src="https://github.com/user-attachments/assets/84841a0e-93e3-456c-9a21-224e7404ff4c" />


Other than that though this looks good to me 👍

Mikefly123 commented:
Hey @hrfarmer I believe you need to remount the CircuitPy drive to allow for the runtime to edit files rather than your computer to edit files. You can read a little more about it in this guide: 

https://learn.adafruit.com/circuitpython-essentials/circuitpython-storage

I don't remember if you can just call `storage.remount("/", readonly=False)` from the REPL, but give it a try and see what happens.

Alandlt15 commented:
Yup @hrfarmer you can try what Michael suggested. How me and @asiemsen  usually do it is by putting that line of code in boot.py. 

Once you do that you wont be able to delete or replace files on the board. So you need to go to the REPL and type these lines to reset the file system:
```
import storage
storage.erase_filesystem()
```
more on that here: https://forums.adafruit.com/viewtopic.php?t=173164

hrfarmer reviewed (changes requested):
Thanks @Mikefly123 @Alandlt15 for the info, was able to get past the read-only error like that. 

However @Alandlt15 I apologize because I misled you earlier, apparently CircuitPython's implementation of the `json` library and `dumps()` [doesn't actually support the indent argument](https://docs.circuitpython.org/en/latest/docs/library/json.html)
<img width="700" alt="image" src="https://github.com/user-attachments/assets/8193c48d-de79-4e14-8454-3434bf4dbab8" />

I tested with removing the argument and it saves to the board & reads the file fine, it just destroys the formatting and writes to a single line. Unless there's another easy way to go about preserving formatting, it'll probably be fine to save without the formatting since users won't be interacting with the config file during runtime.

Alandlt15 commented:
Dam that's a bummer, but you're right I don't think formatting will be an issue. I will remove the indent off the code. Thank you again for looking into this PR!

hrfarmer reviewed (approved):
LGTM!
